vertical-pod-autoscaler/admission-controller/gencerts.sh

```diff
@@ -0,0 +1,78 @@
+#!/bin/bash
+
+# Copyright 2018 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Generates the a CA cert, a server key, and a server cert signed by the CA.
+# reference:
+https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/webhook/gencerts.sh
+set -e
+
+CN_BASE="hpa_webhook"
+
+cat > server.conf << EOF
+[req]
+req_extensions = v3_req
+distinguished_name = req_distinguished_name
+[req_distinguished_name]
+[ v3_req ]
+basicConstraints = CA:FALSE
+keyUsage = nonRepudiation, digitalSignature, keyEncipherment
+extendedKeyUsage = clientAuth, serverAuth
+EOF
+
+# Create a certificate authority
+openssl genrsa -out caKey.pem 2048
+openssl req -x509 -new -nodes -key caKey.pem -days 100000 -out caCert.pem -subj "/CN=${CN_BASE}_ca"
+
+# Create a server certiticate
+openssl genrsa -out serverKey.pem 2048
+# Note the CN is the DNS name of the service of the webhook.
+openssl req -new -key serverKey.pem -out server.csr -subj "/CN=hpa-webhook.kube-system.svc" -config server.conf
+openssl x509 -req -in server.csr -CA caCert.pem -CAkey caKey.pem -CAcreateserial -out serverCert.pem -days 100000 -extensions v3_req -extfile server.conf
+
+outfile=certs.go
+
+cat > $outfile << EOF
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+EOF
+
+echo "// This file was generated using openssl by the gencerts.sh script" >> $outfile
+echo "" >> $outfile
+echo "package main" >> $outfile
+for file in caKey caCert serverKey serverCert; do
+  data=$(cat ${file}.pem)
+  echo "" >> $outfile
+  echo "var $file = []byte(\`$data\`)" >> $outfile
```

Comment:
Embedding certs in image/code is a bad idea. The suggested way of handling certs in K8S is through secrets mechanism. If you are ok with changing certs with every pod restart you can also generate them inside the docker image as a part of the startup script.

Reply:
Went with secrets. The embedded certs were part of the admission webhook example that you pointed me to, but of course you're correct it's not a great production code ;)